### ISSUE TITLE
bug in trainGAM leading to repeated predictors

### ISSUE THREAD

**Issue by ptitle** [adamlilith/enmSdmX#48]
Hi Adam,

I think there is another bug in trainGAM. As another currently open issue points out, we occasionally get an error `"Repeated variables as arguments of a smooth are not permitted"`. The reason for this, I think, is that the way the code sets up interaction term formulas allows for two identical predictors being put together as an interaction. Also, possibly, a pair of predictors can appear twice but in the opposite order. 

The example code in the documentation does not show this as there are only two predictors. But if I do the following, I get the "repeated variables" error. 

```

# environmental rasters
rastFile <- system.file('extdata/madClim.tif', package='enmSdmX')
madClim <- rast(rastFile)

# create some additional dummy variables
madClim[['bio1b']] <- madClim[[1]] * 2
madClim[['bio12b']] <- madClim[[2]] * 5

# coordinate reference system
wgs84 <- getCRS('WGS84')

# lemur occurrence data
data(lemurs)
occs <- lemurs[lemurs$species == 'Eulemur fulvus', ]
occs <- vect(occs, geom=c('longitude', 'latitude'), crs=wgs84)

occs <- elimCellDuplicates(occs, madClim)

occEnv <- extract(madClim, occs, ID = FALSE)
occEnv <- occEnv[complete.cases(occEnv), ]
	
# create 10000 background sites (or as many as raster can support)
bgEnv <- terra::spatSample(madClim, 20000)
bgEnv <- bgEnv[complete.cases(bgEnv), ]
bgEnv <- bgEnv[1:min(10000, nrow(bgEnv)), ]

# collate occurrences and background sites
presBg <- data.frame(
  presBg = c(
    rep(1, nrow(occEnv)),
    rep(0, nrow(bgEnv))
  )
)

env <- rbind(occEnv, bgEnv)
env <- cbind(presBg, env)

predictors <- names(madClim)




ga <- trainGAM(
	data = env,
	resp = 'presBg',
	preds = predictors,
	verbose = TRUE,
	cores = 1
)

Error in { : 
  task 8 failed - "Repeated variables as arguments of a smooth are not permitted"
In addition: Warning message:
In newton(lsp = lsp, X = G$X, y = G$y, Eb = G$Eb, UrS = G$UrS, L = G$L,  :
  Fitting terminated with step failure - check results carefully

```

As a fix, I replaced lines 131-164 with (note this includes a fix for the other interaction bug):

```
# interaction terms
if (length(preds) > 1L & n >= 2 * presPerTermInitial & !is.null(interaction)) {

	predCombos <- expand.grid(list(preds, preds))
	predCombos <- t(apply(predCombos, 1, sort))
	predCombos <- unique(predCombos)
	predCombos <- predCombos[predCombos[, 1] != predCombos[, 2],]
	
	for (i in 1:nrow(predCombos)) {
	
		pred1 <- predCombos[i, 1]
		pred2 <- predCombos[i, 2]

		# create term
		if (!inherits(data[ , pred1], 'factor') & !inherits(data[ , pred2], 'factor')) {

			terms <- c(terms, paste0(interaction, '(', pred1, ', ', pred2, ', bs=\'', smoothingBasis, '\')'))

		} else if (inherits(data[ , pred1], 'factor') & !inherits(data[ , pred2], 'factor')) {

			terms <- c(terms, paste0(interaction, '(', pred2, ', by=', pred1, ', bs=\'', smoothingBasis,'\')'))

		} else if (!inherits(data[ , pred1], 'factor') & inherits(data[ , pred2], 'factor')) {

			terms <- c(terms, paste0(interaction, '(', pred1, ', by=', pred2, ', bs=\'', smoothingBasis, '\')'))

		} else if (inherits(data[ , pred1], 'factor') & inherits(data[ , pred2], 'factor')) {

			terms <- c(terms, paste0(pred1, ' * ', pred2))

		}
	}
				
} # if more than one term

```
And this seems to solve the problem.

**Comment by adamlilith** [comment]
Hi Pascal,
Thank you for catching this, diagnosing it, and fixing it!  I will put the fix into the code base this weekend!
Adam

**Comment by adamlilith** [comment]
Issue fixed in `enmSdmX` 1.1.9, which is now the master and has been pushed to CRAN. Thanks again, Pascal!

**Comment by ptitle** [comment]
FYI, I think the same issue of repeated predictors is happening in trainNS as well.

**Comment by adamlilith** [comment]
Ah, shoot—I even looked through the other trainXYZ() functions to ensure they didn't have the same issue. I'll correct it, but I have to wait a month to resubmit to CRAN.  *Thank you*, as always, for letting me know!

Adam
________________________________

**Comment by adamlilith** [comment]
The issue with `trainNS()` has been fixed in version 1.1.10 (now on the `master` branch). I will push this to CRAN when I am allowed next month. Thanks again, Pascal!
